gitextensions/gitextensions
# Compare artificial commits to all other commits

Compare artificial commits to all other commits

Handled in #4031, based on #4137.
The three commits can be reviewed individually.
This PR is more a review item as it is right now. The three commits could have been submitted together too, it was separated as it has a new test case.

It is possible to compare a normal commit to any other commit, but staged/unstaged has until know just been comparable to its immediate parent
This commit removes that limitation. In order to do that, the diff handling has been cleaned up some in the previous PRs and first commit of this PR. There are maybe not less code related to artificial commits, but the normal flow has less special handling

It is also possible to compare to artificial commits in FormDiff

Unit tests have been added in some other PRs, there are no obvious candidates here.

 
Screenshots before and after (if PR changes UI):
- (staged is parent to unstaged, screenshot 
![image](https://user-images.githubusercontent.com/6248932/32810810-f6bef6ea-c99c-11e7-83ea-efa3ee95dee8.png)


How did I test this code:
 - Select artificial commit
 - Ctrl-select normal commit
 - See the differences

Has been tested on (remove any that don't apply):
 - GIT 2.15 This adds the use of -R option for diff, available since at leastGit 1.7
 - Windows 10
